pkg/cloud/isolated_network.go

```diff
@@ -110,12 +110,24 @@ func (c *client) CreateIsolatedNetwork(fd *infrav1.CloudStackFailureDomain, isoN
 	return c.AddCreatedByCAPCTag(ResourceTypeNetwork, isoNet.Spec.ID)
 }
 
-// OpenFirewallRules opens a CloudStack firewall for an isolated network.
+// OpenFirewallRules opens a CloudStack egress firewall for an isolated network.
 func (c *client) OpenFirewallRules(isoNet *infrav1.CloudStackIsolatedNetwork) (retErr error) {
-	p := c.cs.Firewall.NewCreateEgressFirewallRuleParams(isoNet.Spec.ID, NetworkProtocolTCP)
-	_, retErr = c.cs.Firewall.CreateEgressFirewallRule(p)
-	if retErr != nil && strings.Contains(strings.ToLower(retErr.Error()), "there is already") { // Already a firewall rule here.
-		retErr = nil
+	protocols := []string{NetworkProtocolTCP, NetworkProtocolUDP, NetworkProtocolICMP}
+	for _, proto := range protocols {
+		p := c.cs.Firewall.NewCreateEgressFirewallRuleParams(isoNet.Spec.ID, proto)
+
+		if proto == "icmp" {
+			p.SetIcmptype(-1)
+			p.SetIcmpcode(-1)
+		}
+
+		_, retErr = c.cs.Firewall.CreateEgressFirewallRule(p)
```

Comment:
Do we want to collect the errors into 1 and return that? (We already depend on hashicorps multi error package).

Reply:
Something like this?

```patch
--- a/pkg/cloud/isolated_network.go
+++ b/pkg/cloud/isolated_network.go
@@ -121,12 +121,10 @@ func (c *client) OpenFirewallRules(isoNet *infrav1.CloudStackIsolatedNetwork) (r
 			p.SetIcmpcode(-1)
 		}

-		_, retErr = c.cs.Firewall.CreateEgressFirewallRule(p)
-		if retErr != nil && strings.Contains(strings.ToLower(retErr.Error()), "there is already") { // Already a firewall rule here.
-			retErr = nil
-		}
-		if retErr != nil {
-			break
+		_, err := c.cs.Firewall.CreateEgressFirewallRule(p)
+		if err != nil && !strings.Contains(strings.ToLower(err.Error()), "there is already") { // Already a firewall rule here.
+			multierror.Append(retErr, errors.Wrapf(
+				err, "failed creating egress firewall rule for network ID %s protocol %s", isoNet.Spec.ID, proto))
 		}
 	}
 	c.customMetrics.EvaluateErrorAndIncrementAcsReconciliationErrorCounter(retErr)
 ```

Reply:
Yes, that looks good.